cluster-api/cloud/google/machinesetup/config_types.go

```diff
@@ -0,0 +1,144 @@
+*/
+
+package machinesetup
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"reflect"
+
+	"github.com/ghodss/yaml"
+	clustercommon "k8s.io/kube-deploy/cluster-api/pkg/apis/cluster/common"
+	clusterv1 "k8s.io/kube-deploy/cluster-api/pkg/apis/cluster/v1alpha1"
+)
+
+type ConfigWatch struct {
+	path string
+}
+
+type ValidConfigs struct {
+	configList *configList
+}
+
+type configList struct {
+	Items []config `json:"items"`
+}
+
+type config struct {
+	// A list of the valid combinations of ConfigParams that will
+	// map to the given Image and Metadata.
+	Params []ConfigParams `json:"machineParams"`
+
+	// The fully specified image path.
+	Image    string   `json:"image"`
+	Metadata Metadata `json:"metadata"`
+}
+
+type Metadata struct {
+	StartupScript string `json:"startupScript"`
+}
+
+type ConfigParams struct {
+	OS       string
+	Roles    []clustercommon.MachineRole
+	Versions clusterv1.MachineVersionInfo
+}
+
+func NewConfigWatch(path string) (*ConfigWatch, error) {
+	if _, err := os.Stat(path); err != nil {
+		return nil, err
+	}
+	return &ConfigWatch{path: path}, nil
+}
+
+func (cw *ConfigWatch) ValidConfigs() (*ValidConfigs, error) {
+	file, err := os.Open(cw.path)
+	if err != nil {
+		return nil, err
+	}
+	return parseMachineSetupYaml(file)
+}
+
+func parseMachineSetupYaml(reader io.Reader) (*ValidConfigs, error) {
+	bytes, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+
+	configList := &configList{}
+	err = yaml.Unmarshal(bytes, configList)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ValidConfigs{configList}, nil
+}
+
+func (vc *ValidConfigs) GetYaml() (string, error) {
+	bytes, err := yaml.Marshal(vc.configList)
+	if err != nil {
+		return "", err
+	}
+	return string(bytes), nil
+}
+
+func (vc *ValidConfigs) GetImage(params *ConfigParams) (string, error) {
+	machineSetupConfig, err := vc.matchMachineSetupConfig(params)
+	if err != nil {
+		return "", err
+	}
+	return machineSetupConfig.Image, nil
+}
+
+func (vc *ValidConfigs) GetMetadata(params *ConfigParams) (Metadata, error) {
+	machineSetupConfig, err := vc.matchMachineSetupConfig(params)
+	if err != nil {
+		return Metadata{}, err
+	}
+	return machineSetupConfig.Metadata, nil
+}
+
+func (vc *ValidConfigs) matchMachineSetupConfig(params *ConfigParams) (*config, error) {
+	for _, conf := range vc.configList.Items {
+		for _, validParams := range conf.Params {
+			if params.OS != validParams.OS {
+				continue
+			}
+			validRoles := rolesToMap(validParams.Roles)
+			paramRoles := rolesToMap(params.Roles)
+			if !reflect.DeepEqual(paramRoles, validRoles) {
+				continue
+			}
+			if params.Versions != validParams.Versions {
+				continue
+			}
+			return &conf, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find a matching machine setup config for params %+v", params)
+}
+
+func rolesToMap(roles []clustercommon.MachineRole) map[clustercommon.MachineRole]int {
+	rolesMap := map[clustercommon.MachineRole]int{}
+	for _, role := range roles {
```

Comment:
Yeah it would be a bug or user error if there are duplicate roles. I asked Kris about it yesterday and he said we can add a check for it elsewhere later on.

Reply:
I'm curious why roles was made a slice in the first place. @krousey

Reply:
Others expressed interest in having more than Node and Master roles and also having machines serve multiple roles.